### ISSUE TITLE
Offline background index preparation

### ISSUE THREAD

**Issue by petr-tik** [clangd/clangd#587]
Hey,
First of all thanks for building clangd and opening github issues as a place to submit questions and requests. This might be a question instead of a feature request, if it turns out that you already support this and I couldn't make sense of the documentation. 

## Request/question
I would like to be able to build the clangd background index ahead of time, offline without starting an LSP client to drive the server. 

## Scenario 
I have a mid-sized C++ project (800 kLoC with 100+ debian libraries as external dependencies) that I would like to build a devcontainer for. Considering that containers pretend to give you root when you mount the repository, I was thinking about building the index for the whole project (usually about 400MB) at `docker build` time. That way, I can pull the docker image, start a container and resume from a pre-built and pre-indexed project state. When I start changing my files and recompiling, clangd will only need to compute and analyze the diff between background index and document change events. 

From what I read about clangd, i thought it should be possible. 

## Background
Have been running clangd successfully on several projects (work and home), found it immensely useful. Found `--background-index` and `compile-commands-dir=` flags. Read about clangd-indexer, but found the documentation discouraging its use unless I was a clangd developer, so assumed --background-index should be enough

Decided to try building an offline index on a checkout of mozilla/rr repo on my home desktop without docker at first. 

```bash
git clone git@github.com:mozilla/rr.git
cd rr
mkdir build && cd build
cmake ..
ls -alh compile_commands.json
-rw-rw-r-- 1 petr_tik petr_tik 716K Nov 11 22:45 compile_commands.json
```

## Tried so far
In the root of the repo (rr dir) start clangd with the following arguments. 

```bash
clangd-11 --log=verbose --background-index --pretty -j=24 --compile-
commands-dir=build/
clangd is a language server that provides IDE-like features to editors.

It should be used via an editor plugin rather than invoked directly. For more information, see:
	https://clangd.llvm.org/
	https://microsoft.github.io/language-server-protocol/

clangd accepts flags on the commandline, and in the CLANGD_FLAGS environment variable.

I[22:45:25.804] Ubuntu clangd version 11.0.0-++20201008094836+176249bd673-1~exp1~20201008080315.114
I[22:45:25.804] PID: 106414
I[22:45:25.804] Working directory: /home/petr_tik/Coding/rr
I[22:45:25.804] argv[0]: clangd-11
I[22:45:25.804] argv[1]: --log=verbose
I[22:45:25.804] argv[2]: --background-index
I[22:45:25.804] argv[3]: --pretty
I[22:45:25.804] argv[4]: -j=24
I[22:45:25.804] argv[5]: --compile-commands-dir=build/
V[22:45:25.804] User config file is /home/petr_tik/.config/clangd/config.yaml
I[22:45:25.804] Starting LSP over stdin/stdout
```

### Expected 
An index to start building in either build/ or root of the repo. I used this command to watch for changes and it showed no changes. 

```bash
~/Coding/rr master $
watch -n 1 -d ls -alh build/ .
```

Running lsof on the clangd process didn't show any fds open apart from the standard

```bash
watch -n 1 -d lsof -p $(pgrep clangd)
```

Same result - no changes

```bash
COMMAND      PID     USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
clangd.ma 106414 petr_tik  cwd    DIR  259,2     4096 12460508 /home/petr_tik/Coding/rr
clangd.ma 106414 petr_tik  rtd    DIR  259,1     4096        2 /
clangd.ma 106414 petr_tik  txt    REG  259,1  8639136  1587773 /usr/lib/llvm-11/bin/clangd
clangd.ma 106414 petr_tik  mem    REG  259,1   170784   399415 /lib/x86_64-linux-gnu/libtinfo.so.5.9
clangd.ma 106414 petr_tik  mem    REG  259,1    14560   410145 /lib/x86_64-linux-gnu/libdl-2.27.so
clangd.ma 106414 petr_tik  mem    REG  259,1    31680   410164 /lib/x86_64-linux-gnu/librt-2.27.so
clangd.ma 106414 petr_tik  mem    REG  259,1   116960   399429 /lib/x86_64-linux-gnu/libz.so.1.2.11
clangd.ma 106414 petr_tik  mem    REG  259,1   208080  1452743 /usr/lib/x86_64-linux-gnu/libedit.so.2.0.56
clangd.ma 106414 petr_tik  mem    REG  259,1    31032  1452791 /usr/lib/x86_64-linux-gnu/libffi.so.6.0.4
clangd.ma 106414 petr_tik  mem    REG  259,1  2030544   410141 /lib/x86_64-linux-gnu/libc-2.27.so
clangd.ma 106414 petr_tik  mem    REG  259,1    96616   393908 /lib/x86_64-linux-gnu/libgcc_s.so.1
clangd.ma 106414 petr_tik  mem    REG  259,1  1700792   410146 /lib/x86_64-linux-gnu/libm-2.27.so
clangd.ma 106414 petr_tik  mem    REG  259,1  1594864  1444010 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.25
clangd.ma 106414 petr_tik  mem    REG  259,1 83192616  1443149 /usr/lib/x86_64-linux-gnu/libLLVM-11.so.1
clangd.ma 106414 petr_tik  mem    REG  259,1 46016880  1575175 /usr/lib/llvm-11/lib/libclang-cpp.so.11
clangd.ma 106414 petr_tik  mem    REG  259,1   144976   410161 /lib/x86_64-linux-gnu/libpthread-2.27.so
clangd.ma 106414 petr_tik  mem    REG  259,1   170960   404576 /lib/x86_64-linux-gnu/ld-2.27.so
clangd.ma 106414 petr_tik    0u   CHR  136,3      0t0        6 /dev/pts/3
clangd.ma 106414 petr_tik    1u   CHR  136,3      0t0        6 /dev/pts/3
clangd.ma 106414 petr_tik    2u   CHR  136,3      0t0        6 /dev/pts/3
```

## Questions
1. Do I understand it correctly that the line 
```bash
I[22:45:25.804] Starting LSP over stdin/stdout
```
means it's waiting for an lsp client to connect before it starts building an index?

2. Is it currently possible to build the index offline, as described above? 
3. If so, can I help you document this feature? 
4. Will such an index still be valid when I start an actual interactive session from an LSP client? 
5. If not, do I need clangd-indexer for that or is the outlined usecase currently unsupported?
6. If it's unsupported, do you need anything else from me to consider this issue a feature request? 

Thanks again!

**Comment by kadircet** [comment]
Most of your verdicts are correct. Unfortunately we don't support building background-index without an interactive workflow. You can use clangd-indexer to generate a static-index to provide users of the container with a baseline until background-indexing catches up with the whole project, but the downside is, the static-index built via clangd-indexer cannot be used to provide background-index with a base, so indexing will be performed from scratch (for the first time only).

There's a workaround in https://github.com/clangd/clangd/issues/479#issuecomment-722505550 for mocking an interactive session.

This is enough for a feature request, but we currently don't have the capacity to invest in a solution. In theory it would be nice to have a different run mode in clangd-indexer for producing background-index shards rather than a static-index. So, patches welcome :)

**Comment by petr-tik** [comment]
Thanks for the detailed response and pointing out the workaround. I will give it a try when it becomes a pressing problem. 

> So, patches welcome :)

I like to contribute wherever I can, however, I am v new to C++ (only started doing it in my current job) and I don't want to promise things I cannot deliver. 

> This is enough for a feature request

Great. Happy to provide more detail to anyone who would like to tackle this.

**Comment by petr-tik** [comment]
I’ve been learning more about LSP and found there is a complementary format of storing the index on disk - LSIF. 

https://microsoft.github.io/language-server-protocol/specifications/lsif/0.4.0/specification/

Furthermore, there is already an implementation fork of clangd building an index for non-interactive (read only) query answering. 
https://github.com/sourcegraph/lsif-clang/tree/llvmorg-10.0.0-lsif-clang/src

At the same time, recent commits in main suggest you are working on providing a remote index to answer some queries. I can only guess how useful that could be for an engineering organisation with a mono-repo of Cpp. 

The fork LSIF indexer implementation shows that the serialisation is easy, but what would it take to teach clangd to read the index from a file in LSIF format instead of starting a new indexing process every time it’s started. This would really help in the devcontainer scenario, where containers may be started and stopped relatively frequently, but they can be batch-built ahead of time, where large projects with many dependencies are pre-indexed, so clangd only has to keep track of live changes to the repo in question.

**Comment by kadircet** [comment]
> The fork LSIF indexer implementation shows that the serialisation is easy, so my question is - what would it take to teach clangd to read the index from a file in LSIF format instead of starting a new indexing process every time it’s started?

> This would really help in the devcontainer scenario, where containers may be started and stopped relatively frequently, but they can be batch-built ahead of time, where large projects with many dependencies are pre-indexed, so clangd only has to keep track of live changes to the repo in question.

How is that any different than just building a static index with clangd-indexer and distributing that with your containers? (as i mentioned in my previous comment)

**Comment by petr-tik** [comment]
if I understood this explanation correctly, this is the main difference. 

> You can use clangd-indexer to generate a static-index to provide users of the container with a baseline until background-indexing catches up with the whole project, but the downside is, the static-index built via clangd-indexer cannot be used to provide background-index with a base, so indexing will be performed from scratch (for the first time only).

the fact that these indexes cannot be merged and the fact that using clangd-indexer is discouraged unless i am a developer of clangd, i wanted to share what i found. Given the formation of the LSIF protocol and a live implementation, i was wondering, if that's something that can be considered in the future.

**Comment by sam-mccall** [comment]
LSIF has all the same limitations and more (and is IIUC quite bulky for a full codebase!).
Conceptually LSIF is at the wrong abstraction level to use as an index for clangd, it's closer to exposing whole language-server responses (which are interaction/display-oriented) rather than the semantic info we need.

**Comment by i-ky** [comment]
I've created #1340 for further LSIF discussion.

**Comment by nabaco** [comment]
Bumping this thread, as I'm facing a similar issue.
I'm working in a multi-repo environment, with lots of dependencies between the packages, and I'd like to pre-compile the index shards.
Something along the lines of `clangd-indexer --executor=all-TUs --format=binary-shards compile_commands.json` or `--format=cache` would be nice.
Given that the functionality exists in the clangd binary itself, I think it should be easy to add the functionality for someone that understands the codebase and C++.
Here is the current indexer: https://github.com/llvm/llvm-project/blob/74f207883bc5fe2a7300c4b4f1ff080a107ab148/clang-tools-extra/clangd/indexer/IndexerMain.cpp#L145 (where it explicitly merges the indices)
Here is the merge itself if I understand correctly: https://github.com/llvm/llvm-project/blob/74f207883bc5fe2a7300c4b4f1ff080a107ab148/clang-tools-extra/clangd/indexer/IndexerMain.cpp#L61
And here is the background indexing task in clangd itself: https://github.com/llvm/llvm-project/blob/fc748408c6436c2a80e372156d53e15b014c6105/clang-tools-extra/clangd/index/Background.cpp#L156

**Comment by EgorDuplensky** [comment]
Came across this issue while trying to use mcp-language-server with clangd as a lsp provider.
Please correct me if I am wrong but because of the same limitations it is almost impossible to use clangd as lsp server for MCP protocol.
I.e. it is impossible to find all the symbols that match some name using "workspace/symbol" in interactiveless manner (as an API request from MCP client).

**Comment by HighCommander4** [comment]
@EgorDuplensky why is a static index not suitable for this use case?

**Comment by EgorDuplensky** [comment]
It still does not work right away even with clangd-indexer.
Maybe it is related to this part:
```
FileIndex (“dynamic index”)[#](https://clangd.llvm.org/design/indexing#fileindex-dynamic-index)
This is the top layer, and includes symbols from the files that have been opened and the headers they include. This is used:

to provide code completions for symbols at global scope in header files. (This is more efficient than deserializing big parts of the preamble).
to ensure cross-references for the files you’re working on are available, even if the background index hasn’t finished yet
to ensure locations of definitions/references aren’t stale despite actively editing the file
The FileIndex class stores data from each file separately. When a file is parsed, the TUScheduler invokes a callback which adds the AST to the index. (In fact, there is a separate storage and callback for expensive-and-rare preamble rebuilds vs cheap-and-frequent main-file rebuilds).
```
Even with static index prepared and ready, mcp does not get the result on its first attempt, but it seems to succeed on a second one.

**Comment by HighCommander4** [comment]
> Even with static index prepared and ready, mcp does not get the result on its first attempt, but it seems to succeed on a second one.

On clangd startup, the index (regardless of how it was prepared) needs to be read from disk and loaded into memory, and index serving structures (lookup tables) need to be constructed in memory. LSP requests are not blocked on this process, so it can indeed happen that requests sent very soon after startup are answered without using the index. You can observe this during interactive use as well.

If a client would like to wait for the index to be loaded before sending further requests, I believe clangd sends [`$/progress`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress) notifications that allows you to determine when that has happened.

(I'm happy to answer follow-up questions about this, but please start a new discussion for it, as we are straying from the topic of offline **background** index preparation.)

**Comment by jasonwilliams** [comment]
I would like the option to force-build an index without having to start VS Code and wait for the background threads to do the work. On large repositories like Chromium it can take a long time, and if i have VS  Code open i already need the index built.

I'm not sure if this is the right thread, but is it not possible to have some cli to force build the index and then the LSP can pick up on the database built instead of starting again?

**Comment by HighCommander4** [comment]
> I'm not sure if this is the right thread, but is it not possible to have some cli to force build the index and then the LSP can pick up on the database built instead of starting again?

Yes, I believe that is what's being asked for in this open issue.

Are you interested in contributing an implementation of it?

**Comment by jasonwilliams** [comment]
> > I'm not sure if this is the right thread, but is it not possible to have some cli to force build the index and then the LSP can pick up on the database built instead of starting again?
> 
> Yes, I believe that is what's being asked for in this open issue.
> 
> Are you interested in contributing an implementation of it?

I would be happy to help contribute a fix if someone could point me in the right direction, I am new to this codebase.

Essentially this should just be calling whatever VS Code calls right? But without the background threads argument (I’m sure it’s more complicated than this but that’s the idea I have in my head)

**Comment by jasonwilliams** [comment]
> You can use clangd-indexer to generate a static-index to provide users of the container with a baseline until background-indexing catches up with the whole project, but the downside is, the static-index built via clangd-indexer cannot be used to provide background-index with a base, so indexing will be performed from scratch (for the first time only).

Could someone explain why this is the case? Does it generate a different format or something that the IDE can’t use? I’ve seen some discussion around LSIF, is that what the background indexer generates today?

My use case is a bit different to the OPs but our goal is the same, I would like to build then trigger a index automatically without having to interact and be there, in fact, when I’m interacting I don’t want indexing to be taking place. It feels an odd choice that it only does full indexing on interaction.

**Comment by HighCommander4** [comment]
> > You can use clangd-indexer to generate a static-index to provide users of the container with a baseline until background-indexing catches up with the whole project, but the downside is, the static-index built via clangd-indexer cannot be used to provide background-index with a base, so indexing will be performed from scratch (for the first time only).
> 
> Could someone explain why this is the case? Does it generate a different format or something that the IDE can’t use?

Yes, the static index and the background index have different formats. The static index is a single file, while the background index is "sharded" into pieces, each representing information from a single source file (and with each shard stored in a separate file). This allows the background index to be built and updated incrementally (at the resolution of one source file at a time).

> I’ve seen some discussion around LSIF, is that what the background indexer generates today?

Neither the static index format nor the background index format is related to LSIF.

**Comment by HighCommander4** [comment]
> Essentially this should just be calling whatever VS Code calls right?

In an interactive editing session, the only thing an editor needs to do to trigger background indexing is to send a `textDocument/didOpen` message for a source file in the project. Clangd reacts to this by looking for a `compile_commands.json` file in the ancestor directories of the file being opened, and upon finding one, will start building an index covering all source files listed in the `compile_commands.json`.

One could write a CLI that does exactly what an editor would do, i.e. launch clangd as a subprocess and communicate with it over LSP including sending the `textDocument/didOpen`. Earlier in the discussion someone already referenced [writing a script](https://github.com/clangd/clangd/issues/479#issuecomment-722505550) that does exactly this. This requires no changes to clangd itself.

That said, the above is more of a workaround, and a proper/nicer fix would be to give `clangd` (or `clangd-indexer`) a CLI option that kicks off generation of a background index directly, with no subprocesses or LSP communication or any related infrastructure involved.

**Comment by jasonwilliams** [comment]
I've created a [PR](https://github.com/llvm/llvm-project/pull/175209) which does exactly that, it offers a CLI option to `clangd-indexer` and that will index the directory generating background shards. Please try it out and review if possible
